### PR TITLE
Implement function to extract empty coverage data from an instrumented file

### PR DESCRIPTION
**Pull request by motiz88**
This is in support of fixing https://github.com/istanbuljs/babel-plugin-istanbul/issues/4. Next step would be to consume this new function in `nyc` when not instrumenting.

This won't build until a `babel-traverse` release is made with https://github.com/babel/babel/pull/4746. PRing early so the code can be reviewed until then.

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

File: src/read-coverage.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        allowReturnOutsideFunction: true,
+        allowSuperOutsideMethod: true,
+        sourceType: "script", // I think ?
+        plugins: ["*"]
+    });
+
+    const varName = genVar(filename);
+    let covPath;
+    traverse(ast, {
+        VariableDeclarator: function (path) {
```
Comment (motiz88) [inline src/read-coverage.js]: 😀 

Just to be more precise here:

We parse the entire file, then stop _traversing_ the AST the moment we find a variable declaration whose name is equal to `genVar`. Then we use `babel-traverse`'s robust understanding of scopes and ability to statically evaluate nodes (much the opposite of `valueToNode` used [here](https://github.com/istanbuljs/istanbul-lib-instrument/blob/0225e8cbe3777193d874076c7dc75f1fd567ccc3/src/visitor.js#L469)) to extract the value of `coverageData` from within that IIFE.

The change to `babel-traverse` merely adds support for evaluating object literals, which was a long-standing `TODO` in Babel and of course is essential here.

**Comment by motiz88** [comment]
So my [plan](https://github.com/istanbuljs/babel-plugin-istanbul/issues/4#issuecomment-254309187) is for `nyc` to read the _compiled_ file. This minimizes duplicated work _and_ the risk of breaking something if for some reason (babel config, version conflicts, the weather...) we don't duplicate it exactly right. That is, we defer entirely to whoever instrumented the file in the first place, and try to assume very little about that step - except that it was done by a version of `visitor.js` we're compatible with, and that the names of "our" variables have not been mangled by e.g. a minifier (this is actually a problem but I'm leaving it open atm).

**Comment by motiz88** [comment]
This actually gives me an idea: How about, instead of looking for the `genVar` hash (computed from `sourceFileName` which we actually might not know when we first look at the file! As in the Babel codebase - source files are in `src/` and transpiled to `lib/`; the hashes are for `src/`) we do this:

1. Have `visitor` insert a magic constant as an extra field in the initial coverage object - say, a salted hash of the current istanbul-lib-instrument major version.
2. Have `read-coverage` search directly for an object expression with this magic constant. This doesn't require any prior knowledge of the source file's location (or looking at the source map if any), and is robust in the face of minification & name mangling and most other transforms that could happen after instrumentation.

**Comment by motiz88** [comment]
This just leaves `nyc`: [`--all`](https://github.com/istanbuljs/nyc/blob/99dbbb3a36c12770686f59fa3ff26fbbdc991e8d/bin/nyc.js#L42) is is implemented using [`addAllFiles()`](https://github.com/istanbuljs/nyc/blob/b7a413a6c369402464db9b231692943f851846d6/index.js#L158), which attempts to grab [`instrumenter.lastFileCoverage()`](https://github.com/istanbuljs/nyc/blob/b7a413a6c369402464db9b231692943f851846d6/index.js#L168). With `instrument: false`, `instrumenter` is the noop instrumenter, [for which `lastFileCoverage()` returns `null`](https://github.com/istanbuljs/nyc/blob/b7a413a6c369402464db9b231692943f851846d6/lib/instrumenters/noop.js#L7). I would actually just use the noop instrumenter as a hook for this functionality - invoke `readInitialCoverage` inside its `instrumentSync()` method and use the result for `lastFileCoverage()`.

I'll sketch out a PR to `nyc` and see what needs to be changed here, if anything. (* ahem MONOREPO ahem * 😉 )

**Comment by bySabi** [comment]
I'm currently working on a new Karma Coverage Reporter plugin base on new Istanbul API, say v1?
Yesterday I was trying to implement the 'includeAllSource'  feature ripping some code from `nyc`. 
 The current flow is:
- build code with `babel-plugin-istanbul`
- parse `.babelrc` and extract include/exclude istanbul plugin file rules
- Instrument this files with empty objects using
```
libInstrumenter.createInstrumenter({
    esModules: true     // Enable instrument ES6 modules
  });
```
Is a little hacky  ... 
I don´t understand everything your doing here but please take in consideration this use-case.

I stop  add `includeAllFiles` feature until this is finnish.

**Comment by motiz88** [comment]
@bySabi sounds like the perfect use case for this. You're doing essentially the duplicated/brittle work I was worried about - having to look at .babelrc, instrumenting twice, etc. With this new feature you should be able to just `readInitialCoverage` from the final transpiled code.

You _would_ probably still need to grab hold of nyc's include/exclude config somehow, to best emulate nyc's part of the task.

**Comment by bySabi** [comment]
Ah! fine this is wonderful @motiz88 ... I´m still trying to figure out the "correct way" to grab hold include/exclude rules settings in order of not duplicate it all over the place.

Parsing include/exclude from babelrc need to take in consideration  `env: {  "test": { ` like  keys. 
Ideally we can have a status object with include/exclude rules, probably this is not posible, or extend `test-exclude` package to handle babelrc parsing or is anything better ...

**Comment by bySabi** [comment]
@motiz88 now all included files will be instrumented by default. Thats right?
Then options like `includeAllFiles` or `--all` not make sense anymore.

**Comment by motiz88** [comment]
No, that's not exactly right. The exact same files will be instrumented as
before, only now `--all` will work as intended with `babel-plugin-istanbul`
and correctly report 0% coverage for source files that are never required
in a coverage test run.

On Oct 18, 2016 1:38 PM, "bySabi Files" <notifications@github.com> wrote:

@motiz88 <https://github.com/motiz88> now all included files will be
instrumented by default. Thats right?
Then options like includeAllFiles or --all not make sense anymore.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/istanbuljs/istanbul-lib-instrument/pull/28#issuecomment-254470524>,
or mute the thread

**Comment by bySabi** [comment]
But babel-plugin-istanbul will instrument not `required` modules always or how?

**Comment by motiz88** [comment]
That depends on how you run the plugin. If you compile an entire directory
with `babel-cli` and enable the plugin, it will instrument all of the files
you compile. (Then `--all` will be able to report on all of them, which
this PR will help accomplish.) But if you rely on a bundler to pull in and
transpile files, you may need to explicitly tell it to process any extra
files not found in your `require()` graph.

However, bundler integrations are out of scope for what I'm personally
doing here and in `nyc`. They certainly should be able to put this new API
to good use, though.

**Comment by bySabi** [comment]
Ok,  I got it now! ... Let this is finnished/merged to see what more is needed to do with bundlers. I will be watching it.

Thanks.

**Comment by motiz88** [comment]
So this essentially works - see [this test repo I made](https://github.com/motiz88/nyc-all-demo) and the caveats on istanbuljs/nyc#420.

As for this specific PR, the new implementation is still not entirely robust - it expects certain local variable names to be used in the coverage initialization function, which won't survive e.g. minification. This can definitely be fixed if we identify the variables by order and/or content shape instead (at least if they're not found by name).

(Note: I've accidentally pulled in #29 but I'm going to leave it in unless you object, @bcoe, on the assumption that #29 will get merged soon anyhow)

**Comment by bcoe** [comment]
` it expects certain local variable names to be used in the coverage initialization function, which won't survive e.g. minification.`

I wonder is there any convention for making that a line should be skipped by a minifier, e.g., `/* skipNextLine */.

This is looking good, are we still waiting on the changes to `babel-traverse` to land, curious why we're seeing the failures.

**Comment by elodszopos** [comment]
Anxiously looking forward to this, boys. Keep up the good work.

**Comment by motiz88** [comment]
@bcoe Yeah, there's a Babel release being worked on right now which should include the updated `babel-traverse` (the PR has already been accepted and merged).

**Comment by bcoe** [comment]
@motiz88 just happened to notice a new babel release go out, does that mean we can rebase this and it will work :D

**Pull request merged**